Support JSON column type in recommender.

amrishlal opened the pull request:
This PR adds support for handling JSON column in the recommender service.

jackjlli commented on pinot-controller/src/main/java/org/apache/pinot/controller/recommender/data/generator/JsonGenerator.java:
```diff
@@ -0,0 +1,61 @@
+  // Length of each key-value pair int the JSON string.
+  static final int DEFAULT_JSON_ELEMENT_LENGTH = 5;
+
+  private int _jsonStringLength = 0;
```
make it final?

amrishlal replied:
Fixed.

jackjlli commented on pinot-controller/src/main/java/org/apache/pinot/controller/recommender/data/generator/JsonGenerator.java:
```diff
@@ -0,0 +1,61 @@
+  static final int DEFAULT_JSON_ELEMENT_LENGTH = 5;
+
+  private int _jsonStringLength = 0;
+  private Random _random;
```
Same here.

amrishlal replied:
Fixed.

jackjlli commented on pinot-controller/src/test/java/org/apache/pinot/controller/recommender/data/generator/JsonGeneratorTest.java:
```diff
@@ -0,0 +1,59 @@
+        "{}".length() + (elementCount - 1) + elementCount * JsonGenerator.DEFAULT_JSON_ELEMENT_LENGTH);
+
+    // Check if json string is valid json.
+    JsonUtils.stringToJsonNode(jsonString);
```
Wrap with try catch block, and make it fail if exception is thrown.

jackjlli commented on pinot-controller/src/test/java/org/apache/pinot/controller/recommender/data/generator/JsonGeneratorTest.java:
```diff
@@ -0,0 +1,59 @@
+  @Test
+  public void testNext()
+      throws IOException {
+    // JsonGenerator generates empty json when size is  less than JsonGenerator.DEFAULT_JSON_ELEMENT_LENGTH)
```
Extra space in the sentence.

amrishlal replied:
Fixed.

jackjlli commented on pinot-controller/src/test/java/org/apache/pinot/controller/recommender/data/generator/JsonGeneratorTest.java:
```diff
@@ -0,0 +1,59 @@
+
+
+public class JsonGeneratorTest {
+  @Test
```
Put an empty line before `@Test`.

amrishlal replied:
Fixed.

jasperjiaguo reviewed (changes requested):
I think the wiring will work and we can configure the average length in the schema-with-metadata. Can we add a test case in `org.apache.pinot.controller.recommender.TestConfigEngine` to test segment generation in an actual recommender run? We can follow the `SegmentSizeRuleInput.json` input.

amrishlal commented:
> Can we add a test case in `org.apache.pinot.controller.recommender.TestConfigEngine` to test segment generation in an actual recommender run? We can follow the `SegmentSizeRuleInput.json` input.

Added JSON column to `SegmentSizeRuleInput.json` and updated test case.

amrishlal commented on pinot-controller/src/test/java/org/apache/pinot/controller/recommender/data/generator/JsonGeneratorTest.java:
```diff
@@ -0,0 +1,68 @@
+    // Number of expected elements in the generated JSON string
+    int elementCount = desiredLength / JsonGenerator.DEFAULT_JSON_ELEMENT_LENGTH;
+
+    // Remove escape characters from jsonString for verification purposes. Escape character were added before comma
+    // since json string is written to a CSV file where comma is used as delimiter.
+    jsonString = StringUtils.remove(jsonString, "\\");
```
This is not specific to JSON. The recommender writes all column values to a CSV data file before converting the data file into a Pinot segment. Multi-level JSON is also ok as long as we escape the commas in JSON so that they don't conflict with CSV comma delimiter.

amrishlal commented on pinot-controller/src/test/resources/recommenderInput/SegmentSizeRuleInput.json:
```diff
@@ -131,7 +139,8 @@
     "select f from tableName where (a=0 and b=1) or c=7 or (d = 7 and e =1)": 2,
-    "select f from tableName where t between 1 and 1000": 2
+    "select f from tableName where t between 1 and 1000": 2,
+    "select q from tableName where JSON_MATCH(q, '\"$.a\" IS NOT NULL')": 1
```
The type of JSON predicate doesn't matter for Recommender. GREATER_THAN, LESS_THAN, is currently not supported in JSON predicate but we support others. More details [here](https://docs.pinot.apache.org/users/user-guide-query/json-queries).

walterddr reviewed (approved):
looks good to me

siddharthteotia commented:
Let's add a simple rule to recommend index in the recommender. It can be as simple as "recommend JSON index on all JSON type columns in the input schema". 

Later we can make the rule more fancy / improvements. The purpose of this PR is to ensure that user experience when using this recommender in our internal provisioning flow does not give bad experience / error prone if JSON column is present in schema. 

cc @jasperjiaguo

amrishlal commented:
> Let's add a simple rule to recommend index in the recommender. It can be as simple as "recommend JSON index on all JSON type columns in the input schema".

Added rule to recommend NoDictionary and JsonIndex on JSON columns.

jasperjiaguo commented on pinot-controller/src/main/java/org/apache/pinot/controller/recommender/io/InputManager.java:
```diff
@@ -203,13 +204,15 @@ private void reorderDimsAndBuildMap()
     Set<String> noDictionaryColumns = _overWrittenConfigs.getIndexConfig().getNoDictionaryColumns();
 
     /*Validate if there's conflict between NoDictionaryColumns and dimNamesWithAnyIndex*/
     Set<String> dimNamesWithAnyIndex = new HashSet<>();
     dimNamesWithAnyIndex.add(sortedColumn);
     dimNamesWithAnyIndex.addAll(invertedIndexColumns);
     dimNamesWithAnyIndex.addAll(rangeIndexColumns);
+    dimNamesWithAnyIndex.addAll(jsonIndexColumns);
```
This part is essentially to ensure sortedColumn/invertedIndexColumns/rangeIndexColumns are dictionary encoded. On the contrary, json index must be raw. L215 needs to be removed. Maybe we should rename `dimNamesWithAnyIndex` to `dimNamesWithDictionaryDependentIndex`.

jasperjiaguo replied:
https://github.com/apache/pinot/pull/8988
